### PR TITLE
feat (Typeahead, ModalPicker): Configurable page size

### PR DESCRIPTION
**Pull request by arjanvlek**
Added the ability to specify the number of suggestions for
`TypeaheadSingle` and `TypeaheadMultiple`. Also added the ability to
show all predefined options matching the input. One of our applications
needs to show as much suggestions as possible, as users need to know
which values are supported.

Added the ability to change the page size of `ModalPickerMultiple` and
`ModalPickerSingle`, for some applications it is desired to have more
than 10 items per page on screen.

### CODE REVIEW

File: src/form/Typeahead/single/TypeaheadSingle.tsx [hunk 1]
```diff
@@ -67,7 +83,10 @@ export default function TypeaheadSingle<T>(props: Props<T>) {
     reloadOptions,
     query,
     pageNumber: 1,
-    size: 10,
+    size:
+      Array.isArray(options) && showAllSuggestionsInDropdown
```
Comment (MrHus) [inline src/form/Typeahead/single/TypeaheadSingle.tsx:87]: The `showAllSuggestionsInDropdown` is not necessary take a look at `RadioGroup`, `Select` and `CheckboxMultipleSelect` they simply when the options is an array use the length of the array as the size.

File: src/form/ModalPicker/multiple/ModalPickerMultiple.tsx [hunk 2]
```diff
@@ -49,6 +49,12 @@ export type Props<T> = Omit<
      */
     canSearch?: boolean;
 
+    /**
+     * Optionally specify the number of elements to show per page in the modal.
```
Comment (MrHus) [inline src/form/ModalPicker/multiple/ModalPickerMultiple.tsx:53]: I Question if we really want to be able do do this. Fetching many options will be a performance hit, and showing to many options will make the modal too large which is detrimental to the user experience. 

Can you give me an example of when this is useful in your application?

File: src/form/Typeahead/single/TypeaheadSingle.tsx (hunk 1 above)
Comment (arjanvlek) [reply src/form/Typeahead/single/TypeaheadSingle.tsx]: I have removed the `showAllSuggestionsInDropdown` prop with the latest patch.

File: src/form/ModalPicker/multiple/ModalPickerMultiple.tsx (hunk 2 above)
Comment (arjanvlek) [reply src/form/ModalPicker/multiple/ModalPickerMultiple.tsx]: While I agree that fetching too many options degrades performance and UX, a customer of ours has requested to increase the options slightly above 10. 

As for the typeahead, we are using predefined options, so performance is not a concern.

File: src/form/ModalPicker/single/ModalPickerSingle.stories.tsx [hunk 3]
```diff
@@ -41,6 +41,26 @@ storiesOf('Form/ModalPicker/ModalPickerSingle', module)
...
+          placeholder="Please select your province"
+          options={provinces()}
+          labelForOption={(province) => province.label}
+          value={value}
+          onChange={setValue}
+          pageSize={5}
+        />
+
+        {value ? <p>Your chosen province is: {value.label}</p> : null}
+      </Form>
```
Comment (gidomanders) [inline src/form/ModalPicker/single/ModalPickerSingle.stories.tsx:61]: Please also add a note about high page size deteriorating UX on smaller screens (same note as in the property comments)

File: src/form/Typeahead/multiple/TypeaheadMultiple.stories.tsx [hunk 4]
```diff
@@ -69,6 +69,34 @@ storiesOf('Form/Typeahead/JarbTypeaheadMultiple', module)
...
+          pageSize={5}
+        />
+
+        {value ? (
+          <p>
+            Your chosen provinces are:{' '}
+            {value.map((province) => province.label).join(', ')}
+          </p>
+        ) : null}
+      </Form>
```
Comment (gidomanders) [inline src/form/Typeahead/multiple/TypeaheadMultiple.stories.tsx:97]: Here a similar note about high page size

File: src/form/Typeahead/multiple/TypeaheadMultiple.tsx [hunk 5]
```diff
@@ -21,7 +21,24 @@ import { useId } from '../../../hooks/useId/useId';
...
+     * in the dropdown.
+     *
+     * When `options` is an array, all options will always be shown
+     * and this prop has no effect.
+     *
+     * When `options` is a fetcher, it will determine how many options
+     * are requested through the fetcher as the `page` parameter.
+     * This means that when you set the pageSize to `100` that
+     * `100` items are fetched from the back-end. Beware of
+     * performance issues when setting the value too high.
```
Comment (gidomanders) [inline src/form/Typeahead/multiple/TypeaheadMultiple.tsx:36]: Please add the UX deterioration here too to make the comment exactly the same on all components. Maybe this property could even be moved to the FieldCompatibleWithPredeterminedOptions type?

File: src/form/Typeahead/single/TypeaheadSingle.stories.tsx [hunk 6]
```diff
@@ -55,6 +55,28 @@ storiesOf('Form/Typeahead/JarbTypeaheadSingle', module)
...
+          placeholder="Please select your province"
+          options={provinceFetcher}
+          labelForOption={(province) => province.label}
+          value={value}
+          onChange={setValue}
+          pageSize={5}
+        />
+
+        {value ? <p>Your chosen province is: {value.label}</p> : null}
+      </Form>
```
Comment (gidomanders) [inline src/form/Typeahead/single/TypeaheadSingle.stories.tsx:77]: Please add the aforementioned note here too

File: src/form/Typeahead/single/TypeaheadSingle.test.tsx [hunk 7]
```diff
@@ -90,6 +93,34 @@ describe('Component: TypeaheadSingle', () => {
...
+
+      expect(toJson(typeaheadSingle)).toMatchSnapshot(
+        'Component: TypeaheadSingle => ui => async with a custom pageSize of 2 options in the dropdown'
+      );
+    });
+
+    test('async with the default pageSize of 10 options in the dropdown', () => {
+      const { typeaheadSingle } = setup({
+        isAsync: true,
+        pageSize: undefined
```
Comment (gidomanders) [inline src/form/Typeahead/single/TypeaheadSingle.test.tsx:113]: You don't have to specify undefined parameters as the default is undefined

File: src/form/Typeahead/single/TypeaheadSingle.tsx [hunk 8]
```diff
@@ -17,7 +17,24 @@ import { optionToTypeaheadOption } from '../utils';
...
+     * in the dropdown.
+     *
+     * When `options` is an array, all options will always be shown
+     * and this prop has no effect.
+     *
+     * When `options` is a fetcher it will determine how many options
+     * are requested through the fetcher as the `page` parameter.
+     * This means that when you set the pageSize to `100` that
+     * `100` items are fetched from the back-end. Beware of
+     * performance issues when setting the value too high.
```
Comment (gidomanders) [inline src/form/Typeahead/single/TypeaheadSingle.tsx:32]: Same here about the deteriorating UX

**Review by gidomanders — CHANGES_REQUESTED** [review]

**Review by MrHus — APPROVED** [review]

**Review by gidomanders — APPROVED** [review]

**Pull request merged**